### ISSUE TITLE
Anaconda (linux) fails to import SWIG interfaced C++ code when std_vector.i is included

### ISSUE THREAD

**Issue by montes00** [ContinuumIO/anaconda-issues#1172]
I have faced a problem on my C++ program binding to Anaconda.
When I tried to use the `std_vector.i`, import in Anaconda fails (undefined symbol).

Measurements I have seen are:
- The sample below represents the bug reported here. It does not work in Anaconda Python for Linux (64 bit, python3.5).
- However, the identical code works for python3 installed by the apt package system in Ubuntu 16.04. (The same machine)
- The identical code works in Anaconda Python for MacOSX (64 bit, python3.5).

The code below represents the bug. It essentially is the same as in http://www.swig.org/Doc3.0/Library.html#Library_std_vector
- Install swig via conda command.   conda install conda
- Create a header file (C++)

```
/* File : example.h */

#include <vector>
#include <algorithm>
#include <functional>
#include <numeric>

double average(std::vector<int> v) {
    return std::accumulate(v.begin(),v.end(),0.0)/v.size();
}
```
- Create a SWIG interface file (example.i)

```
%module example
%{
#include "example.h"
%}

%include "std_vector.i"
// Instantiate templates used by example
namespace std {
   %template(IntVector) vector<int>;
   %template(DoubleVector) vector<double>;
}

// Include the header file with above prototypes
%include "example.h"
```
- Run SWIG: swig -c++ -python example.i
- Compile via setup.py: python setup.py build_ext --inplace

```
from distutils.core import setup,Extension

example_module = Extension('_example',
        sources=['example_wrap.cxx'],
        )

setup(name='example',
        version='0.1',
        ext_modules=[example_module],
        )
```
- Execute.  python -c 'import example'

For Anaconda for Linux, I face the problem.

```
Traceback (most recent call last):
  File "/home/user/tmp/simple_test/vector_i/example.py", line 18, in swig_import_helper
    return importlib.import_module(mname)
  File "/home/user/anaconda3/envs/irfpy.sara-develop/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 666, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 577, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 906, in create_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
ImportError: /home/user/tmp/simple_test/vector_i/_example.cpython-35m-x86_64-linux-gnu.so: undefined symbol: _ZNSt12out_of_rangeC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/user/tmp/simple_test/vector_i/example.py", line 21, in <module>
    _example = swig_import_helper()
  File "/home/user/tmp/simple_test/vector_i/example.py", line 20, in swig_import_helper
    return importlib.import_module('_example')
  File "/home/user/anaconda3/envs/irfpy.sara-develop/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ImportError: /home/user/tmp/simple_test/vector_i/_example.cpython-35m-x86_64-linux-gnu.so: undefined symbol: _ZNSt12out_of_rangeC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
```

Current conda install is

```
Current conda install:

               platform : linux-64
          conda version : 4.2.9
       conda is private : False
      conda-env version : 4.2.9
    conda-build version : 2.0.2
         python version : 3.5.2.final.0
       requests version : 2.11.1
       root environment : /home/user/anaconda3  (writable)
    default environment : /home/user/anaconda3/envs/swig_test_vector
       envs directories : /home/user/anaconda3/envs
          package cache : /home/user/anaconda3/pkgs
           channel URLs : https://repo.continuum.io/pkgs/free/linux-64/
                          https://repo.continuum.io/pkgs/free/noarch/
                          https://repo.continuum.io/pkgs/pro/linux-64/
                          https://repo.continuum.io/pkgs/pro/noarch/
            config file : None
           offline mode : False
```

**Comment by nehaljwani** [comment]
Closing as unable to reproduce this on linux-64 with`swig-3.0.10-0`. Feel free to re-open if you stiil face the same issue.

**Comment by samuelstjean** [comment]
I have the same problem here, it builds and fail on import. I was installing with pip as outlined below a prebuilt package with swig (by myself, some time ago), and of course it used to work before, anaconda is up to date also

 pip install --force https://github.com/samuelstjean/spams-python/releases/download/0.1/spams-python-v2.5-python3.zip -U

~~~python
samuel ~ $ conda update anaconda
Fetching package metadata .........
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /home/samuel/anaconda3:
#
anaconda                  4.3.1               np111py35_0 
~~~

Error itself from the callee software (also stuff I did, so pretty sure it works)

~~~python
samuel ~/Téléchargements $ nlsam_denoising -h
Traceback (most recent call last):
  File "/home/samuel/python/nlsam/denoiser.py", line 18, in <module>
    import spams
  File "/home/samuel/anaconda3/lib/python3.5/site-packages/spams.py", line 6, in <module>
    import spams_wrap
  File "/home/samuel/anaconda3/lib/python3.5/site-packages/spams_wrap.py", line 24, in <module>
    _spams_wrap = swig_import_helper()
  File "/home/samuel/anaconda3/lib/python3.5/site-packages/spams_wrap.py", line 23, in swig_import_helper
    return importlib.import_module(mname)
  File "/home/samuel/anaconda3/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ImportError: /home/samuel/anaconda3/lib/python3.5/site-packages/_spams_wrap.cpython-35m-x86_64-linux-gnu.so: undefined symbol: _ZTVNSt7__cxx1119basic_istringstreamIcSt11char_traitsIcESaIcEEE

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/samuel/anaconda3/bin/nlsam_denoising", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/samuel/git/nlsam/scripts/nlsam_denoising", line 23, in <module>
    from nlsam.denoiser import nlsam_denoise
  File "/home/samuel/python/nlsam/denoiser.py", line 20, in <module>
    raise ImportError("Couldn't find spams library, is the package correctly installed?")
ImportError: Couldn't find spams library, is the package correctly installed?
~~~

As this was recently closed, how did it get fixed/could not get reproduced? Newer version of swig maybe?

**Comment by nehaljwani** [comment]
@samuelstjean I am not sure whether your issue is related to `swig`. Does the following work for you?

```
conda create -yn 1172 -c conda-forge python-spams
source activate 1172
python -c 'import spams'
```

**Comment by samuelstjean** [comment]
I actually fixed it on my side by rebuilding from the swig source wrapper instead of using the conda forge one.

I may have an old version of gcc in anaconda because I made the mistake of installing a package from a random repo, and they always tend to change your list of installed stuff and break everything not long afterwards.

For the record, can a recent system gcc interfere with package from anaconda/anaconda gcc or that is not a thing? As I run archlinux, anaconda will most likely always lag behind.

Unfortunately people who come up with rather similar error (like me) ight end up here as this is still one of the first post to show up on google.
